s2fft/utils/jax_pritimive.py

```diff
@@ -0,0 +1,136 @@
+from abc import ABCMeta, abstractmethod
+from dataclasses import dataclass
+from functools import partial
+from typing import Tuple
+
+import jax.numpy as jnp
+from jax import jit, lax
+from jax._src.api import ShapeDtypeStruct
+from jax._src.core import ShapedArray
+from jax._src.typing import Array, ArrayLike
+from jax.experimental.custom_partitioning import custom_partitioning
+from jax.lax import dynamic_slice
+from jax.sharding import Mesh, NamedSharding
+from jax.sharding import PartitionSpec as P
+from abc import ABCMeta, abstractmethod
+from dataclasses import dataclass
+from typing import Tuple, Sequence, Union, Callable
+from functools import partial, reduce
+import operator
+import os
+import warnings
+
+import numpy as np
+import jax.numpy as jnp
+from jax.lib import xla_client
+from jax import core, dtypes
+from jax.interpreters import xla, mlir
+from jax.experimental.custom_partitioning import custom_partitioning
+from jax.interpreters.mlir import ir, dtype_to_ir_type
+from jax._src.interpreters import batching
+from jax._src import dispatch
+
+# Inspired by https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/jax/cpp_extensions.py
+# and by the tweaked version in https://github.com/DifferentiableUniverseInitiative/jaxDecomp/blob/main/jaxdecomp/_src/spmd_ops.py
```

Comment:
The issue here is that we can't include a derivative version of the code from `NVIDIA/TransformerEngine` repository which is under [the Apache 2.0 License](https://github.com/NVIDIA/TransformerEngine/blob/release_v1.7/LICENSE) unless we meet the conditions of redistribution under the terms of that license, which in particular require us to

>       (a) You must give any other recipients of the Work or
>           Derivative Works a copy of this License; and
> 
>       (b) You must cause any modified files to carry prominent notices
>           stating that You changed the files; and
> 
>       (c) You must retain, in the Source form of any Derivative Works
>           that You distribute, all copyright, patent, trademark, and
>           attribution notices from the Source form of the Work,
>           excluding those notices that do not pertain to any part of
>           the Derivative Works; and

So here we either need to 

  1. include a copy of the original license, a notice stating we modified the file and an attribution to the original copyright  holder (NVIDIA),
  2. or remove this code.

I would favour the latter, because I don't think there is much advantage to having an abstract `BasePrimitive` class here. We could get similar code modularity just with a function something like

```Python
def register_primitive(
    name: str,
    multiple_results: bool,
    abstract_evaluation: Callable,
    lowering_per_platform: dict[None | str, Callable],
    batcher: None | Callable = None,
    jacobian_vector_product: None | Callable = None,
    transpose: None | Callable = None,
):
    primitive = core.Primitive(name)
    primitive.multiple_results = multiple_results
    primitive.def_impl(partial(xla.apply_primitive, primitive))
    primitive.def_abstract_eval(abstract_evaluation)
    for platform, lowering in lowering_per_platform.items():
        mlir.register_lowering(primitive, lowering, platform=platform)
    if batcher is not None:
        batching.primitive_batchers[primitive] = batcher
    if jacobian_vector_product is not None:
        ad.primitive_jvps[primitive] = jacobian_vector_product
    if transpose is not None:
        ad.primitive_transposes[primitive] = transpose
    return primitive
```